src/librbd/api/Mirror.cc

```diff
@@ -3144,96 +3148,42 @@ int Mirror<I>::group_promote(IoCtx& group_ioctx, const char *group_name,
-        if (!delta_images.empty()) {
-          for (auto &s : delta_images) {
-            if (std::find(rollback_images.begin(), rollback_images.end(), s)
-                != rollback_images.end()) {
-              removed_images.push_back(s);
-            } else {
-              inserted_images.push_back(s);
-            }
-          }
-        }
-        if (!removed_images.empty()) {
-          // insert back to membership
-          for (auto &s : removed_images) {
-            IoCtx image_ioctx;
-            r = librbd::util::create_ioctx(group_ioctx, "image", s.pool_id, {},
-                                           &image_ioctx);
-            if (r < 0) {
-              return r;
-            }
-
-            r = Group<I>::image_add(group_ioctx, group_name,
-                                    image_ioctx, s.image_id.c_str());
-            if (r < 0) {
-              lderr(cct) << "error inserting image to group: " << group_name
-                         << " :" << cpp_strerror(r) << dendl;
-              return r;
-            }
-          }
-        }
-
-        if(!inserted_images.empty()) {
-          // remove from the membershiip
-          for (auto &s : inserted_images) {
-            IoCtx image_ioctx;
-            r = librbd::util::create_ioctx(group_ioctx, "image", s.pool_id, {},
-                                           &image_ioctx);
-            if (r < 0) {
-              return r;
-            }
-            r = Group<I>::image_remove(group_ioctx, group_name,
-                                       image_ioctx, s.image_id.c_str());
-            if (r < 0 && r != -ENOENT) {
-              lderr(cct) << "error removing image from a group: " << group_name
-                         << " :" << cpp_strerror(r) << dendl;
-              return r;
-            }
-          }
-        }
-      } // Fixing membership done.
+    if (snap == snaps.rend()) {
+      lderr(cct) << "failed to rollback, no mirror group snapshot available"
+        << dendl;
+      return -EINVAL;
+    }
+    // Check for group membership match
+    std::vector<cls::rbd::GroupImageSpec> rollback_images;
+    for (auto& it : snap->snaps) {
+      rollback_images.emplace_back(it.image_id, it.pool);
+    }
 
-      librbd::NoOpProgressContext prog_ctx;
-      r = Group<I>::snap_rollback(group_ioctx,
-                                  group_name, snap->name.c_str(), prog_ctx);
-      if (r < 0) {
-        lderr(cct) << "failed to rollback to group snapshot: " << snap->name
-                   << " :" << cpp_strerror(r) << dendl;
-        return r;
-      }
-      ldout(cct, 5) << "successfully rollbacked to group snapshot: "
-                    << snap->name << dendl;
-      // Rollback to last good snapshot done
-    } else {
-      lderr(cct) << "failed to rollback, no initial group snapshot available"
+    if (rollback_images != current_images) {
+      lderr(cct) << "group snapshot membership does not match group membership"
```

Comment:
Include the rollback_snap_id in the message - the user will not know which snapshot was selected as the rollback.